pennylane_qiskit/qiskit_device2.py

```diff
@@ -494,37 +472,54 @@ def execute_circuits(session):
 
     def _execute_sampler(self, circuit, session):
         """Execution for the Sampler primitive"""
+        qcirc = [circuit_to_qiskit(circuit, self.num_wires, diagonalize=True, measure=True)]
+        sampler = Sampler(session=session)
+        compiled_circuits = self.compile_circuits(qcirc)
+        sampler.options.update(**self._kwargs)
+
+        # len(compiled_circuits) is always 1 so the indexing does not matter.
+        result = sampler.run(
+            compiled_circuits,
+            shots=circuit.shots.total_shots if circuit.shots.total_shots else None,
+        ).result()[0]
+        classical_register_name = compiled_circuits[0].cregs[0].name
+        self._current_job = getattr(result.data, classical_register_name)
 
-        qcirc = circuit_to_qiskit(circuit, self.num_wires, diagonalize=True, measure=True)
-        if circuit.shots:
-            self.options.execution.shots = circuit.shots.total_shots
-        sampler = Sampler(session=session, options=self.options)
-
-        result = sampler.run(qcirc).result()
-        self._current_job = result
+        results = []
 
         # needs processing function to convert to the correct format for states, and
         # also handle instances where wires were specified in probs, and for multiple probs measurements
         # single_measurement = len(circuit.measurements) == 1
         # res = (res[0], ) if single_measurement else tuple(res)
 
-        return (result.quasi_dists[0],)
+        self._samples = self.generate_samples(0)
+        res = [
+            mp.process_samples(self._samples, wire_order=self.wires) for mp in circuit.measurements
+        ]
+        single_measurement = len(circuit.measurements) == 1
+        res = res[0] if single_measurement else tuple(res)
+        results.append(res)
+
+        return tuple(results)
```

Comment:
This looks like it adds an unexpected dimension to the results for the non-`single_measurement case`. I think we want to go straight to `tuple(res)`.

If so, that makes we think maybe we don't have tests covering the case where `len(res) != 1`? So maybe we should also add a test that would have noticed this.

Reply:
We don't have tests covering the case where `len(res) != 1` 😅 and you're right that it does in fact add an additional layer of dimensionality. Interestingly enough though this will pass any test where `len(res) != 1` regardless, because of the `reorder_fn(res)` function. Basically we would pass in a result that looks like ((x, y),) rather than (x, y)... but then 

`return result[0] if len(result) == 1 else result` will interpret that as result[0] and we're right back to (x, y) -> so the extra layer of dimensionality would be cancelled out. Tricky bug and I'm actually thinking about reworking this whole post-processing the result with `reorder_fn` thing cause 1. it's confusing, and 2. it messes with the Tracker. But I fixed the bug for now and I'll add a few more test cases.

Reply:
This still seems off to me. It looks like for the len(res) != 1 case, we've just gone from :

```
>>> result = []
>>> res = tuple([1, 2])
>>> result.append(res)
>>> tuple(result)
((1, 2),)
```

to 

```
>>> result = []
>>> res = [1, 2]
>>> result.append(res)
>>> tuple(result)
([1, 2],)
```

when I think what we generally expect is:

```
>>> result = []
>>> res = [1, 2]
>>> tuple(res)
(1, 2)
```
But maybe I'm mixing things up here

Reply:
I guess as long as it comes out as expected in the end, it doesn't matter 😅

Reply:
Essentially when the measurementProcess is something like expval() it doesn't work as expected but when it's something like samples() then it will behave more normally.